include/nbl/builtin/hlsl/emulated/float64_t.hlsl

```diff
@@ -0,0 +1,657 @@
+                        resultMantissa >>= 1;
+                    }
+                }
+
+                uint64_t resultBiasedExp = uint64_t(exp) + ieee754::traits<float64_t>::exponentBias;
+                resultMantissa &= ieee754::traits<float64_t>::mantissaMask;
+                uint64_t output = emulated_float64_t_impl::assembleFloat64(lhsSign, resultBiasedExp << ieee754::traits<float64_t>::mantissaBitCnt, resultMantissa);
+                return bit_cast<this_t>(output);
+            }
+
+            // not implemented
+            if (!FlushDenormToZero)
+                return bit_cast<this_t>(0xdeadbeefbadcaffeull);
+        }
+
+        emulated_float64_t operator+(float rhs)
+        {
+            return bit_cast<this_t>(data) + create(rhs);
+        }
+
+        emulated_float64_t operator-(emulated_float64_t rhs) NBL_CONST_MEMBER_FUNC
+        {
+            emulated_float64_t lhs = bit_cast<this_t>(data);
+            emulated_float64_t rhsFlipped = rhs.flipSign();
+            
+            return lhs + rhsFlipped;
+        }
+
+        emulated_float64_t operator-(float rhs) NBL_CONST_MEMBER_FUNC
+        {
+            return bit_cast<this_t>(data) - create(rhs);
+        }
+
+        emulated_float64_t operator*(emulated_float64_t rhs) NBL_CONST_MEMBER_FUNC
+        {
+            // TODO: remove
+            float64_t sum = bit_cast<float64_t>(data) * bit_cast<float64_t>(rhs.data);
+            uint64_t sumAsUint = bit_cast<uint64_t>(sum);
+
+            this_t output2;
+            output2.data = sumAsUint;
+
+            return output2;
+
+            if(FlushDenormToZero)
+            {
+                emulated_float64_t retval = this_t::create(0ull);
+
+                int lhsBiasedExp = ieee754::extractBiasedExponent(data);
+                int rhsBiasedExp = ieee754::extractBiasedExponent(rhs.data);
+
+                uint64_t lhsData = emulated_float64_t_impl::flushDenormToZero(lhsBiasedExp, data);
+                uint64_t rhsData = emulated_float64_t_impl::flushDenormToZero(rhsBiasedExp, rhs.data);
+
+                uint64_t lhsSign = lhsData & ieee754::traits<float64_t>::signMask;
+                uint64_t rhsSign = rhsData & ieee754::traits<float64_t>::signMask;
+                uint64_t sign = (lhsData ^ rhsData) & ieee754::traits<float64_t>::signMask;
+
+                uint64_t lhsMantissa = ieee754::extractMantissa(lhsData);
+                uint64_t rhsMantissa = ieee754::extractMantissa(rhsData);
+
+                int exp = int(lhsBiasedExp + rhsBiasedExp) - ieee754::traits<float64_t>::exponentBias;
+                if (!FastMath)
+                {
+                    if (tgmath::isNaN(lhsData) || tgmath::isNaN(rhsData))
+                        return bit_cast<this_t>(ieee754::traits<float64_t>::quietNaN | sign);
+                    if (tgmath::isInf(lhsData) || tgmath::isInf(rhsData))
+                        return bit_cast<this_t>(ieee754::traits<float64_t>::inf | sign);
```

Comment:
why is this block of code so far down, I'd move it all the way up, so we can assume afterwards that neither input is INF, NaN, or denormalized or zero (Depending on the flush setting)